design/proposals/dbt-delete-merge-schema-branch.md

```diff
@@ -0,0 +1,66 @@
+# Add DBT delete and merge branch-schema operations
+
+---
+
+## Background
+
+To use dbt, users must have an already initialized schema under their dbt profile's target.    
+When using dbt alongside lakeFS, the schema above should have a branch's URI as its location.
+When the users want to branch out and run dbt over a different branch, they first need to create a branch to work on and then create a dbt schema dedicated for that branch (that might have the same name of the branch [by default] or a custom one). #2988 added an option to use `lakectl dbt create-branch-schema --create-branch` to create the dedicated branch and schema at once.
+
+After running dbt over that branch and conducting tests and validations we would want to either: 
+- delete the branch and the schema associated with it
+- merge them into master/main/production/other branch
+- both.
+
+## Goals 
+- Enable deletion of schemas created for dbt branches using lakectl.
+- Enable merge of schemas created for dbt branches using lakectl.
+
+## Non-goals
+
+- Couple branch delete and schema delete operations.
+- Couple branch merge and schema merge operations.
+
+## Suggestions
+
+- `delete-branch-schema` will allow deleting a schema based on a branch name.
+- `merge-branch-schema` will allow merging a schema to a destination schema by branch names.
+
+### `delete-branch-schema` Usage
+
+`lakectl dbt delete-branch-schema --branch <branch name>`: Delete the schema associated with the input branch name.
+* Example: `lakectl dbt delete-branch-schema --branch dbt_jon` will delete the schema associated with the `dbt_jon` branch.
+
+### `merge-branch-schema` Usage
+
+`lakectl dbt merge-branch-schema --source-branch <source branch name> --destination-branch <destination branch name>`: Merge the source schema (associated with the source branch name) to the destination schema (associated with the destination branch name)
+* Example: `lakectl dbt merge-branch-schema --source-branch dbt_jon --destination-branch master` will merge the schema associated with the `dbt_jon` branch to the schema associated with the `master` branch.
+
+## Required changes
+
+In order to implement the new commands, there's a need to map branches to schemas.
+This is necessary because the branch and the schema might have different names (when adding the `--to-schema <custom name>` flag when creating the schema using the `lakectl dbt create-branch-schema` command) and not use the default behavior (which is the same name for the branch and schema).
```

Comment:
Can we get rid of this?  (E.g. clarify requirements by asking users.)

The fundamental issue here is that we have 2 different objects (lakeFS branch, DBT schema) which must be coordinated but are separately managed.  To be very clear: this is not caused in any way by this proposal, or indeed by the entire `lakectl dbt` feature -- it is just the way things are.  At the very least these objects must have a 1:1 mapping!  So any feature which allows this 1:1 property to fail can have really unpleasant consequences.

One possible nasty consequence:
1. Associate branch `main` with `working-schema`.
2. Work on `main` and on the schema.
3. Branch out of `main` to `ariels-dev`, then map it separately to `working-schema` (this is a mistake).
4. Delete branch `ariels-dev` because I made a mistake.
   * `lakectl` deletes `working-schema`.

This is pretty much my fault, but it would be entirely avoidable if schema names were identical to branch names.  (It is quite hard to avoid this programmatically by checking for re-use, because that assumes we know of all accesses to schema file contents.  E.g. if I restore some DBT files from backup, any checks made at time of creation are no longer valid.)

Reply:
I agree that this flag might have a very limited use and therefore can be omitted, but even if so, it leaves us with the case where the user wants to merge back to the main branch (or whatever branch that was associated with the dbt target schema initially). We have no record of which schema is associated with that branch. We can impose a requirement to name the schema as the branch being used, but it sounds a bit too restrictive to me.